src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1758 @@
+                'but specifying the manual implementation will be required from Transformers version v5.0.0 onwards. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+            )
+            return super().forward(
+                hidden_states=hidden_states,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                past_key_value=past_key_value,
+                output_attentions=output_attentions,
+                use_cache=use_cache,
+            )
+
+        bsz, q_len, _ = hidden_states.size()
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        if past_key_value is not None:
+            key_states, value_states = past_key_value.update(key_states, value_states, layer_idx)
+
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        causal_mask = attention_mask
+        if attention_mask is not None:
+            causal_mask = causal_mask[:, :, :, : key_states.shape[-2]]
+
+        # SDPA with memory-efficient backend is currently (torch==2.1.2) bugged with non-contiguous inputs with custom attn_mask,
+        # Reference: https://github.com/pytorch/pytorch/issues/112577.
+        if query_states.device.type == "cuda" and attention_mask is not None:
+            query_states = query_states.contiguous()
+            key_states = key_states.contiguous()
+            value_states = value_states.contiguous()
+
+        softmax_scale = 1 / (query_states.shape[-1] / 2) ** 0.5
+
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=causal_mask,
+            dropout_p=self.attention_dropout if self.training else 0.0,
+            # The q_len > 1 is necessary to match with AttentionMaskConverter.to_causal_4d that does not create a causal mask in case q_len == 1.
+            is_causal=self.is_causal and attention_mask is None and q_len > 1,
+            scale=softmax_scale,
+        )
+
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.view(bsz, q_len, 2 * self.hidden_size)
+
+        attn_output = self.o_proj(attn_output)
+
+        return attn_output, None, past_key_value
+
+
+ZAMBA_ATTENTION_CLASSES = {
+    "eager": ZambaAttention,
+    "flash_attention_2": ZambaFlashAttention2,
+    "sdpa": ZambaSdpaAttention,
+}
+
+
+class ZambaMambaMixer(nn.Module):
```

Comment:
could you help me and write here the diff with the JambaMixer?

Reply:
I find it fairly similar to the MambaMixer otherwise

Reply:
We added a description of differences with the original Mamba layer here https://github.com/Zyphra/transformers_zamba/commit/4ab88a29e8b17696f1b013d4f9d68a5e59fcf83d#diff-0f4d89960530c068a10af906f1958ed46e3e5f2ff937d6be61517478f383b074R511:

```
This module differs from `transformers.models.mamba.modeling_mamba.MambaMixer` in two ways:
    - Added multi-head: the output of `self.in_proj` is split into `self.n_mamba_heads` heads, and each head
    undergoes an independent forward pass, identical to the original `MambaMixer`, up until the pre-activations of
    `self.out_proj`. The pre-activations, coming from different mamba heads, are then concatenated and fed into `self.out_proj`.
    - Added `attention_mask` for batched inference: this tensor multiplies input and output of the convolution layer, setting 
    to zero embeddings associated with `attention_mask == 0` thus preventing the layer to attend to such tokens.
```

Reply:
- multi-head is supported already in `codestral`, but that's `mamba2` so okay probably different
- attention mask support was shipped to `mamba` so should be irrelevant now

Reply:
ok!